File: java/GeigerApiPluginConnector/src/test/java/eu/cybergeiger/api/TestSerialization.java
package eu.cybergeiger.api;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.UUID;

import eu.cybergeiger.api.message.GeigerUrl;
import eu.cybergeiger.api.message.Message;
import eu.cybergeiger.api.message.MessageType;
import eu.cybergeiger.api.plugin.CommunicationSecret;
import eu.cybergeiger.api.plugin.Declaration;
import eu.cybergeiger.api.plugin.MenuItem;
import eu.cybergeiger.api.plugin.PluginInformation;
import eu.cybergeiger.api.utils.StorableHashMap;
import eu.cybergeiger.storage.Visibility;
import eu.cybergeiger.storage.node.DefaultNode;
import eu.cybergeiger.storage.node.Node;
import eu.cybergeiger.storage.node.value.DefaultNodeValue;
import eu.cybergeiger.storage.node.value.NodeValue;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * <p>Unit testing of serializable objects.</p>
 */
public class TestSerialization {


  /**
   * <p>Tests the serialization of the GeigerUrl object.</p>
   */
  @Test
  public void geigerUrlSerializationTest() throws IOException {
    GeigerUrl p = new GeigerUrl("id", "path");
    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    p.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    GeigerUrl p2 = GeigerUrl.fromByteArrayStream(bin);
    assertThat(p).isEqualTo(p2);
  }

  /**
   * <p>Tests the serialization of the MenuItem object.</p>
   */
  @Test
  public void menuItemSerializationTest() throws IOException {
    MenuItem p = new MenuItem(
      new DefaultNode(
        ":menu",
        "plugin",
        Visibility.RED,
        new NodeValue[]{
          new DefaultNodeValue(MenuItem.NAME_KEY, "test"),
          new DefaultNodeValue(MenuItem.TOOLTIP_KEY, "test")
        },
        new Node[0]
      ),
      new GeigerUrl("plugin", "menu")
    );

    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    p.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    MenuItem p2 = MenuItem.fromByteArrayStream(bin);
    assertThat(p).isEqualTo(p2);
  }

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  @Test
  public void pluginInformationSerializationTest() throws IOException {
    PluginInformation p = new PluginInformation("", "exec", 1234, Declaration.DO_NOT_SHARE_DATA);
    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    p.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    PluginInformation p2 = PluginInformation.fromByteArrayStream(bin);
    PluginInformation p3 = PluginInformation.fromByteArray(p.toByteArray());
    assertThat(p.hashCode())
      .isEqualTo(p2.hashCode())
      .isEqualTo(p3.hashCode());
  }

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  @Test
  public void communicationSecretSerializationTest() throws IOException {
    CommunicationSecret[] comsec = new CommunicationSecret[]{
      new CommunicationSecret("Hello".getBytes(StandardCharsets.UTF_8)),
      new CommunicationSecret()
    };

    for (CommunicationSecret p : comsec) {
      ByteArrayOutputStream bout = new ByteArrayOutputStream();
      p.toByteArrayStream(bout);
      ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
      CommunicationSecret p2 = CommunicationSecret.fromByteArrayStream(bin);
      assertThat(p.toString()).isEqualTo(p2.toString());
    }
  }

  /**
   * <p>Tests the serialization of the a serializable Hashmap object.</p>
   */
  @Test
  public void storableMapSerializationTest() throws IOException {
    StorableHashMap hm = new StorableHashMap();
    ByteArrayOutputStream bout = new ByteArrayOutputStream();
    hm.toByteArrayStream(bout);
    ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
    StorableHashMap hm2 = new StorableHashMap();
    StorableHashMap.fromByteArrayStream(bin, hm2);
    assertThat(hm).isEqualTo(hm2);
  }

  /**
   * <p>Tests the serialization of the a serializable Hashmap object.</p>
   */
  @Test
  public void messageSerializationTest() throws IOException {
    for (Message m : new Message[]{
      new Message("src", "target", MessageType.DEACTIVATE_PLUGIN, new GeigerUrl("geiger://id1/path1")),
      new Message("src", "target", MessageType.DEACTIVATE_PLUGIN, null),
      new Message("src", "target", MessageType.DEACTIVATE_PLUGIN, new GeigerUrl("id", "path2")),
      new Message("src", "target", MessageType.DEACTIVATE_PLUGIN, new GeigerUrl("id1", "path")),
    }) {
      ByteArrayOutputStream bout = new ByteArrayOutputStream();
      m.toByteArrayStream(bout);
      ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
      Message m2 = Message.fromByteArrayStream(bin);
      assertThat(m).isEqualTo(m2);
    }
  }
}

File: dart/geiger_api/test/test_serialization.dart
library geiger_api;

import 'dart:async';

import 'package:geiger_api/geiger_api.dart';
import 'package:geiger_api/src/plugin/communication_secret.dart';
import 'package:geiger_localstorage/geiger_localstorage.dart';
import 'package:test/expect.dart';
import 'package:test/scaffolding.dart';
import 'package:uuid/uuid.dart';

import 'print_logger.dart';

class TestSerialization {}

/// Unit testing of serializable objects.
void main() {
  printLogger();
  /**
   * <p>Tests the serialization of the GeigerUrl object.</p>
   */
  test('geigerUrlSerializationTest', () async {
    GeigerUrl p = GeigerUrl(null, GeigerApi.masterId, 'path');
    ByteSink bout = ByteSink();
    p.toByteArrayStream(bout);
    bout.close();
    ByteStream bin = ByteStream(Stream<List<int>>.value(await bout.bytes));

    /// TODO(unassigned): What is that below?
    await GeigerUrl.fromByteArrayStream(bin);

    expect(p.toString() == p.toString(), true,
        reason: 'Cloned GeigerUrls are unequal');
  });

  /**
   * <p>Tests the serialization of the MenuItem object.</p>
   */
  test('menuItemSerializationTest', () async {
    final MenuItem p1 = MenuItem(
        await NodeImpl.fromPath(':menu:1111-1111-111111-111113:testMenu', 'pid',
            nodeValues: [NodeValueImpl('name', 'menuTest')]),
        GeigerUrl(null, GeigerApi.masterId, 'testMenu'),
        true);

    ByteSink bout = ByteSink();
    p1.toByteArrayStream(bout);
    bout.close();
    ByteStream bin = ByteStream(null, await bout.bytes);
    MenuItem p2 = await MenuItem.fromByteArrayStream(bin);
    expect(p1.toString() == p2.toString(), true,
        reason: 'Cloned MenuItems are unequal');
  });

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  test('pluginInformationSerializationTest', () async {
    PluginInformation p = PluginInformation('plugin1', 'exec', 1234,
        Declaration.doNotShareData, CommunicationSecret('Hello'.codeUnits));
    ByteSink bout = ByteSink();
    p.toByteArrayStream(bout);
    bout.close();
    ByteStream bin = ByteStream(Stream<List<int>>.value(await bout.bytes));
    PluginInformation p2 = await PluginInformation.fromByteArrayStream(bin);
    PluginInformation p3 =
        await PluginInformation.fromByteArray(await p.toByteArray()!);
    expect(p.hashCode == p2.hashCode, true,
        reason: 'Cloned Plugininformation using stream are not equal');
    expect(p.hashCode == p3.hashCode, true,
        reason: 'Cloned Plugininformation using array are not equal');
  });

  /**
   * <p>Tests the serialization of the ParameterList object.</p>
   */
  test('communcationSecretSerialization', () async {
    List<CommunicationSecret> comsec = [
      CommunicationSecret('hello'.codeUnits),
      CommunicationSecret('a'.codeUnits),
      CommunicationSecret([35])
    ];

    for (CommunicationSecret p in comsec) {
      ByteSink bout = ByteSink();
      p.toByteArrayStream(bout);
      bout.close();
      ByteStream bin = ByteStream(null, await bout.bytes);
      CommunicationSecret p2 =
          await CommunicationSecret.fromByteArrayStream(bin);
      expect(p.toString() == p2.toString(), true,
          reason: 'Cloned Plugininformation using stream are not equal');
    }
  });

  /**
   * <p>Tests the serialization of the a serializable Hashmap object.</p>
   */
  test('storableMapSerializationTest', () async {
    StorableHashMap<GeigerUrl, Serializer> hm =
        StorableHashMap<GeigerUrl, Serializer>();
    GeigerUrl url = GeigerUrl(null, GeigerApi.masterId, 'path');
    Map<GeigerUrl, Serializer> it = <GeigerUrl, Serializer>{
      url: url,
      url: const StorableString('string')
    };
    hm.addAll(it);
    ByteSink bout = ByteSink();
    hm.toByteArrayStream(bout);
    bout.close();
    ByteStream bin = ByteStream(null, await bout.bytes);
    StorableHashMap hm2 = StorableHashMap();
    await StorableHashMap.fromByteArrayStream(bin, hm2);
    expect(hm2.toString() == hm.toString(), true,
        reason: 'Cloned PluginInformation using stream are not equal');
  });

  /**
   * <p>Tests the serialization of the a serializable Hashmap object.</p>
   */
  test('messageSerializationTest', () async {
    List<Message> messageList = [
      Message('src', 'target', MessageType.deactivatePlugin,
          GeigerUrl(null, GeigerApi.masterId, 'geiger://id1/path1'), null),
      Message('src', 'target', MessageType.deactivatePlugin, null, []),
      Message('src', 'target', MessageType.deactivatePlugin,
          GeigerUrl(null, GeigerApi.masterId, 'path2'), []),
      Message('src', 'target', MessageType.deactivatePlugin,
          GeigerUrl(null, GeigerApi.masterId, ''), <int>[]),
      Message('src', 'target', MessageType.deactivatePlugin,
          GeigerUrl(null, GeigerApi.masterId, ''), <int>[]),
      Message('src', 'target', MessageType.allEvents, null, <int>[]),
      Message('src', 'target', MessageType.allEvents, null,
          const Uuid().v4().codeUnits)
    ];

    for (Message m in messageList) {
      ByteSink bout = ByteSink();
      m.toByteArrayStream(bout);
      bout.close();
      ByteStream bin = ByteStream(Stream<List<int>>.value(await bout.bytes));
      Message m2 = await Message.fromByteArray(bin);
      expect(m == m2, true,
          reason: 'Cloned Plugininformation using stream are not equal');
    }
  });
}
